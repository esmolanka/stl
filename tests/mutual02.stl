
type Delay (a : Type) = < Now : { head : a, tail : Stream } | Later : Delay >
  with Stream = < Yield : Delay | Skip : Stream | EndOfStream >

#eval Delay Int

#eval Stream Int

#check Delay Int <: < Later : Delay Int | Now : {head : Int, tail : Stream Int} >
