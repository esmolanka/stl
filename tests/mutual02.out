tests/mutual02.types:5:1-5:16:
  Normalised:
    μ Delay. ∀ (ρ : Row).
      < Now? :
          ∃ (ρ : Row).
            { head : Int
            , tail :
                μ Stream. ∀ (ρ : Row).
                  <Yield? : Delay | Skip? : Stream | EndOfStream? : Unit | … ρ>
            , … ρ
            }
      | Later? : Delay
      | … ρ
      >
  Kind:
    Type
tests/mutual02.types:7:1-7:17:
  Normalised:
    μ Stream. ∀ (ρ : Row).
      < Yield? :
          μ Delay. ∀ (ρ : Row).
            < Now? : ∃ (ρ : Row). {head : Int, tail : Stream, … ρ}
            | Later? : Delay
            | … ρ
            >
      | Skip? : Stream
      | EndOfStream? : Unit
      | … ρ
      >
  Kind:
    Type
tests/mutual02.types:9:1-9:82:
  Subtype:
    μ Delay. ∀ (ρ : Row).
      < Now? :
          ∃ (ρ : Row).
            { head : Int
            , tail :
                μ Stream. ∀ (ρ : Row).
                  <Yield? : Delay | Skip? : Stream | EndOfStream? : Unit | … ρ>
            , … ρ
            }
      | Later? : Delay
      | … ρ
      >
  Supertype:
    ∀ (ρ : Row).
      < Later? :
          μ Delay. ∀ (ρ : Row).
            < Now? :
                ∃ (ρ : Row).
                  { head : Int
                  , tail :
                      μ Stream. ∀ (ρ : Row).
                        < Yield? : Delay
                        | Skip? : Stream
                        | EndOfStream? : Unit
                        | … ρ
                        >
                  , … ρ
                  }
            | Later? : Delay
            | … ρ
            >
      | Now? :
          ∃ (ρ : Row).
            { head : Int
            , tail :
                μ Stream. ∀ (ρ : Row).
                  < Yield? :
                      μ Delay. ∀ (ρ : Row).
                        < Now? : ∃ (ρ : Row). {head : Int, tail : Stream, … ρ}
                        | Later? : Delay
                        | … ρ
                        >
                  | Skip? : Stream
                  | EndOfStream? : Unit
                  | … ρ
                  >
            , … ρ
            }
      | … ρ
      >
  Result:
    OK